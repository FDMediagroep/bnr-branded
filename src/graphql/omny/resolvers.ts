import { Program } from '../../utils/models';
import {
    getClipDetails,
    getClipDetailsExt,
    getClipTranscript,
    getPlaylistClips,
    getPlaylistDetails,
    getProgramClips,
    getProgramDetails,
    getProgramPlaylists,
    getPrograms,
} from '../../utils/omnyHelper';

async function ProgramsResolver(_, req) {
    return (await getPrograms(process.env.OMNY_ORGID))?.Programs ?? [];
}

async function ProgramResolver(_, req) {
    return await getProgramDetails(
        process.env.OMNY_ORGID,
        _?.ProgramId ?? req?.programId
    );
}

async function ProgramClipsResolver(_: Program, req) {
    return await getProgramClips(
        process.env.OMNY_ORGID,
        _?.Id ?? req?.programId,
        req?.cursor,
        req?.pageSize
    );
}

async function ProgramPlaylistResolver(_, req) {
    return await getProgramPlaylists(
        process.env.OMNY_ORGID,
        _?.Id ?? req?.programId
    );
}

async function PlaylistDetailsResolver(_, req) {
    return await getPlaylistDetails(process.env.OMNY_ORGID, req?.playlistId);
}

async function PlaylistClipsResolver(_, req) {
    return await getPlaylistClips(
        process.env.OMNY_ORGID,
        _?.Id ?? req?.playlistId,
        req?.cursor,
        req?.pageSize
    );
}

async function ClipResolver(_, req) {
    return await getClipDetails(process.env.OMNY_ORGID, req?.clipId);
}

async function ClipExternalResolver(_, req) {
    return await getClipDetailsExt(process.env.OMNY_ORGID, req?.externalId);
}

async function ClipTranscriptResolver(_, req) {
    return await getClipTranscript(
        process.env.OMNY_ORGID,
        req?.clipId,
        req?.format,
        req?.speakers
    );
}

export const OmnyQueryResolvers = {
    programs: ProgramsResolver,
    program: ProgramResolver,
    programClips: ProgramClipsResolver,
    playlists: ProgramPlaylistResolver,
    playlist: PlaylistDetailsResolver,
    playlistClips: PlaylistClipsResolver,
    clip: ClipResolver,
    clipExternal: ClipExternalResolver,
    clipTranscript: ClipTranscriptResolver,
};

/**
 * Here starts the N+1 problems for GraphQL.
 */
export const OmnyFieldResolvers = {
    /**
     * type `Program` special resolvers for
     * specific fields.
     */
    Program: {
        Clips: ProgramClipsResolver,
        Playlists: ProgramPlaylistResolver,
    },
    Clip: {
        Program: ProgramResolver,
    },
    Playlist: {
        Program: ProgramResolver,
        PlaylistClips: PlaylistClipsResolver,
    },
};
