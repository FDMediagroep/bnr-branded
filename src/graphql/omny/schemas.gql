"""
OmnyStudio Program type
"""
type Program {
    """
    The ID of the program.
    """
    Id: String
    """
    The name of the program.
    """
    Name: String
    """
    The URL slug of the program. The Omny.fm show page can be accessed in the format https://omny.fm/shows/{Slug}.
    """
    Slug: String
    """
    The description of the program.
    """
    Description: String
    DescriptionHtml: String
    Author: String
    Copyright: String
    """
    The publisher of the program.
    """
    Publisher: String
    """
    A publicly-accessible URL to the artwork image for this program. The size parameter specifies the image size (default size is medium). Valid size parameters are: thumbnail (64x64), small (256x256), medium (600x600) and large (3000x3000).
    """
    ArtworkUrl: String
    """
    The category of the program.
    """
    Category: String
    Categories: [String]
    """
    The web URL of the program.
    """
    SocialWeb: String
    """
    The Twitter account username of the program (should be accessed in the format https://twitter.com/{SocialTwitter}).
    """
    SocialTwitter: String
    """
    The Facebook page username of the program (should be accessed in the format https://fb.com/{SocialFacebook}).
    """
    SocialFacebook: String
    """
    If true, the program should be considered "private" and should not be shown in public directories or indexed. (Private programs are not available in the consumer API).
    """
    Hidden: Boolean
    """
    If true, the program should be considered an archive and will not be actively updated.
    """
    Archived: Boolean
    """
    The name of the network group of the program.
    """
    Network: String
    NetworkId: String
    ExternalId: String
    ContactName: String
    ContactEmail: String
    DefaultPlaylistId: String
    """
    This field is convenient but also causes N+1 problem.
    Because if you query this field it will perform an extra data-fetch using
    data needed from the current fetch, usually the ID.
    """
    Clips(cursor: Int, pageSize: Int): Clips
    """
    This field is convenient but also causes N+1 problem.
    Because if you query this field it will perform an extra data-fetch using
    data needed from the current fetch, usually the ID.
    """
    Playlists(cursor: Int, pageSize: Int): Playlists
}

type DirectoryLinks {
    """
    The Apple Podcasts directory URL.
    """
    ApplePodcasts: String
    """
    The Google Podcasts directory URL.
    """
    GooglePodcasts: String
    """
    The Spotify directory URL.
    """
    Spotify: String
    """
    The Stitcher directory URL.
    """
    Stitcher: String
    """
    The TuneIn directory URL.
    """
    TuneIn: String
    """
    (String, obsolete) The Google Play directory URL.
    """
    GooglePlay: String
    """
    The RSS feed URL for this playlist.
    """
    RssFeed: String
}

"""
Get the playlists for an Omny Studio program. Only publicly-listed playlists (public) are shown;
unlisted or private playlists won't appear in the response.
"""
type Playlist {
    """
    The ID of the playlist.
    """
    Id: String
    """
    The name of the playlist.
    """
    Title: String
    """
    The description of the playlist.
    """
    Description: String
    """
    The ID of the program to which the playlist belongs.
    """
    ProgramId: String
    """
    The RSS feed URL for this playlist.
    """
    RssFeedUrl: String
    """
    The embed player iframe URL for this playlist.
    """
    EmbedUrl: String
    """
    A publicly-accessible URL to the artwork image for this playlist. The size parameter specifies the image size (default size is medium). Valid size parameters are: thumbnail (64x64), small (256x256), medium (600x600) and large (3000x3000).
    """
    ArtworkUrl: String
    """
    The number of clips available in the playlist
    """
    NumberOfClips: Int
    """
    The visibility of the clip. Valid values are Public, Unlisted and Private.
    """
    Visibility: String
    """
    The list of iTunes category names.
    """
    Categories: [String]
    """
    The podcast directory (Apple Podcasts, Google Podcasts, Spotify) URLs for this playlist.
    """
    DirectoryLinks: DirectoryLinks
    """
    This field is convenient but also causes N+1 problem.
    Because if you query this field it will perform an extra data-fetch using
    data needed from the current fetch, usually the ID.
    """
    Program: Program
    """
    This field is convenient but also causes N+1 problem.
    Because if you query this field it will perform an extra data-fetch using
    data needed from the current fetch, usually the ID.
    """
    PlaylistClips(cursor: Int, pageSize: Int): Clips
}

type Playlists {
    Playlists: [Playlist]
    Cursor: Int
    pageSize: Int
}

type ClipChapter {
    """
    The ID of the chapter.
    """
    Id: String
    """
    The name of the chapter.
    """
    Name: String
    """
    The timestamp of the chapter in hh:mm:ss.
    """
    Position: String
}

type ClipMonetization {
    """
    If pre-roll ad is enabled for this clip.
    """
    PreRoll: String
    """
    If post-roll ad is enabled for this clip.
    """
    PostRoll: String
    """
    (TimeSpan[]) An array of midroll locations. For example, a value of ["00:00:30", "01:23:45"] would represent a midroll at 30 seconds, and one at 1 hour, 23 minutes and 45 seconds into the episode.
    """
    MidRolls: [String]
}

type RecordingMetadata {
    """
    The date and time the recording started capturing.
    """
    CaptureStartUtc: String
    """
    The date and time the recording finished capturing.
    """
    CaptureEndUtc: String
}

type Clip {
    Id: String
    """
    The title of the clip.
    """
    Title: String
    """
    The description of the clip (without any HTML formatting).
    """
    Description: String
    """
    The description of the clip (with HTML formatting).
    """
    DescriptionHtml: String
    """
    An array of tags of the clips.
    """
    Tags: [String]
    """
    An optional season Number for the clip (as defined by Apple Podcasts).
    """
    Season: Int
    """
    An optional episode Number for the clip (as defined by Apple Podcasts).
    """
    Episode: Int
    """
    The type of episode of the clip (as defined by Apple Podcasts). Valid values are Full, Trailer and Bonus.
    """
    EpisodeType: String
    """
    A publicly-accessible URL to the artwork image for this clip. The size parameter specifies the image size (default size is medium). Valid size parameters are: thumbnail (64x64), small (256x256), medium (600x600) and large (3000x3000).
    """
    ImageUrl: String
    """
    A publicly-accessible URL to the MP3 audio for this clip.
    """
    AudioUrl: String
    """
    If a visualized video has been generated, a publicly-accessible URL to the MP4 video for this clip.
    """
    VideoUrl: String
    """
    The embed player iframe URL for this clip.
    """
    EmbedUrl: String
    """
    The time length of the clip in seconds.
    """
    DurationSeconds: Float
    """
    The publish state of the clip. Clips should have a state of Published.
    """
    PublishState: String
    """
    A publicly-accessible URL for the clip's Omny.fm web player
    """
    PublishedUrl: String
    """
    The visibility state of the clip. Valid visibilities are Public (should be publicly shown in directories and indexes), and Unlisted (should not be shown in directories and indexes).
    """
    Visibility: String
    """
    The date and time when the clip was published.
    """
    PublishedUtc: String
    """
    An array of IDs for playlists to which the clip has been added.
    """
    PlaylistIds: [String]
    """
    An array of chapter objects added to the clip.
    """
    Chapters: [ClipChapter]
    """
    The state of the clip. Valid states are: Ready (the clip is ready to be played).
    """
    State: String
    """
    An optional override for the share URL.
    """
    ShareUrl: String
    """
    An optional episode identifier if the clip was imported from another provider.
    """
    ImportId: String
    """
    Monetization settings for the clip.
    """
    Monetization: ClipMonetization
    """
    Related information about the recording associated with the clip if made from a recording.
    """
    RecordingMetadata: RecordingMetadata
    """
    The ID of the program the to which the clip belongs.
    """
    ProgramId: String
    """
    If the clip is published, the size of the audio file in bytes.
    """
    PublishedAudioSizeInBytes: Int
    """
    The content rating of the of the clip. Valid ratings are Unrated, Clean and Explicit.
    """
    ContentRating: String
    """
    The external ID of the clip.
    """
    ExternalId: String
    Program: Program
}

type Clips {
    Clips: [Clip]
    Cursor: Int
    TotalCount: Int
}

type Query {
    """
    Get the programs for an Omny Studio organization.
    """
    programs: [Program]

    """
    Get the metadata for an Omny Studio program.
    * \`programId\` The ID of the Omny Studio program.
    """
    program(programId: String!): Program

    """
    Get the clips for an Omny Studio program. Only publicly-listed clips (public) are shown;
    unlisted or private clips won't appear in the response.
    * \`programId\` The ID of the Omny Studio program.
    * \`cursor\` (Optional) cursor The paging cursor value to use in order to fetch the next page.
    * \`pageSize\` (Optional) pageSize The number of clips to return in one page.
    """
    programClips(programId: String!, cursor: Int = 1, pageSize: Int = 10): Clips

    """
    Get the playlists for an Omny Studio program. Only publicly-listed playlists (public) are shown;
    unlisted or private playlists won't appear in the response.
    * \`programId\` The ID of the Omny Studio program.
    """
    playlists(programId: String!): Playlists

    """
    Get the metadata for an Omny Studio playlist (which can also be considered a podcast).
    * \`playlistId\` The ID of the Omny Studio playlist.
    """
    playlist(playlistId: String!): Playlist

    """
    Get the clips for an Omny Studio playlist. Only publicly-available clips in the playlist
    (public & unlisted) are shown; private clips won't appear in the response.
    * \`playlistId\` The ID of the Omny Studio playlist.
    * \`cursor\` (Optional) cursor The paging cursor value to use in order to fetch the next page.
    * \`pageSize\` (Optional) pageSize The number of clips to return in one page.
    """
    playlistClips(
        playlistId: String!
        cursor: Int = 1
        pageSize: Int = 10
    ): Clips

    """
    Get the metadata for an Omny Studio clip (episode).
    * \`clipId\` The ID of the Omny Studio clip.
    """
    clip(clipId: String!): Clip

    """
    Get the metadata for an Omny Studio clip (episode) by external ID.
    * \`externalId\` The external ID of the Omny Studio clip (URL encoded).
    """
    clipExternal(externalId: String!): Clip

    """
    Get the transcript for an Omny Studio clip (episode).
    * \`clipId\` The ID of the Omny Studio clip.
    * \`format\` (Optional) format The format in which the transcript will be exported (default is JSON ). Valid parameters are: SubRip, WebVTT, Text, TextWithTimestamps and JSON.
    * \`speakers\` (Optional) speakers Whether a transcript will include speaker names – or default speaker names if speaker names haven't been assigned (default is true). Valid parameters are true and false.
    """
    clipTranscript(
        clipId: String!
        format: String = "JSON"
        speakers: Boolean = true
    ): Clip
}
