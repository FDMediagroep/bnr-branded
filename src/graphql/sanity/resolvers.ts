import { getProgramEnrichment } from '../../utils/sanityHelper';

async function ProgramEnrichmentsResolver(_, req) {
    console.log(_?.Id ?? req?.programId);
    return await getProgramEnrichment(_?.Id ?? req?.programId);
}

export const sanityResolvers = {
    Query: {
        programEnrichments: ProgramEnrichmentsResolver,
    },
    Program: {
        ProgramEnrichments: ProgramEnrichmentsResolver,
    },
};
