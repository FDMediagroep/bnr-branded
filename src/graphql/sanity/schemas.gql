type Program {
    """
    Below field is convenient but also causes N+1 problem.
    Because if you query this field it will perform an extra data-fetch using
    data needed from the current fetch, usually the ID.
    """
    ProgramEnrichments: [ProgramEnrichment]
}

"""
Description of a sponsor of a program
"""
type ProgramSponsor {
    """
    URL to sponsor
    """
    url: String
    """
    Name of sponsor
    """
    name: String
    """
    Logo of sponsor
    """
    logo: String
}

"""
Program enrichment from Sanity CMS.
"""
type ProgramEnrichment {
    """
    Program color
    """
    color: String
    """
    Program sponsors
    """
    sponsors: [ProgramSponsor]
}

type Query {
    """
    Get the program enrichment from Sanity
    """
    programEnrichments(programId: String!): [ProgramEnrichment]
}
