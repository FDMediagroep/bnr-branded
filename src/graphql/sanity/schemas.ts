const ProgramSchema = `
type Program {
    """
    Below field is convenient but also causes N+1 problem.
    Because if you query this field it will perform an extra data-fetch using
    data needed from the current fetch, usually the ID.
    """
    ProgramEnrichments: [ProgramEnrichment]
}`;
const ProgramSponsorSchema = `
"""
Description of a sponsor of a program
"""
type ProgramSponsor {
    """URL to sponsor"""
    url: String
    """Name of sponsor"""
    name: String
    """Logo of sponsor"""
    logo: String
}`;

const ProgramEnrichmentSchema = `
"""
Program enrichment from Sanity CMS.
"""
type ProgramEnrichment {
    """Program color"""
    color: String
    """Program sponsors"""
    Sponsors: [ProgramSponsor]
}`;

/**
 * For GraphQL `type Query`
 */
const ProgramEnrichmentsQuery = `
 """
 Get the program enrichment from Sanity
 """
 programEnrichments(programId: String!): [ProgramEnrichment]
 `;

/**
 * typeDefs
 */
export const sanityTypeDefs = `
${ProgramSponsorSchema}
${ProgramEnrichmentSchema}
${ProgramSchema}

type Query {
    ${ProgramEnrichmentsQuery}
}
 `;
