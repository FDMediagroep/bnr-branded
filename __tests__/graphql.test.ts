import { makeExecutableSchema } from '@graphql-tools/schema';
import { mergeResolvers, mergeTypeDefs } from '@graphql-tools/merge';
import { omnyTypeDefs } from '../src/graphql/omny/schemas';
import { sanityTypeDefs } from '../src/graphql/sanity/schemas';
import { omnyResolvers } from '../src/graphql/omny/resolvers';
import { sanityResolvers } from '../src/graphql/sanity/resolvers';

describe('GraphQL', () => {
    const typeDefs = mergeTypeDefs([omnyTypeDefs, sanityTypeDefs]);
    const resolvers = mergeResolvers([omnyResolvers, sanityResolvers]);
    const executableSchema = makeExecutableSchema({
        typeDefs,
        resolvers,
    });

    test('schema is backward compatible', () => {
        expect(executableSchema).toMatchSnapshot();
    });
});
