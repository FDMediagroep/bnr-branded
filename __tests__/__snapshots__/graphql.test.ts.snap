// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQL schema is backward compatible 1`] = `
GraphQLSchema {
  "__validationErrors": undefined,
  "_directives": Array [
    "@include",
    "@skip",
    "@deprecated",
    "@specifiedBy",
  ],
  "_implementationsMap": Object {},
  "_mutationType": undefined,
  "_queryType": "Query",
  "_subTypeMap": Object {},
  "_subscriptionType": undefined,
  "_typeMap": Object {
    "Boolean": "Boolean",
    "Clip": "Clip",
    "ClipChapter": "ClipChapter",
    "ClipMonetization": "ClipMonetization",
    "Clips": "Clips",
    "DirectoryLinks": "DirectoryLinks",
    "Float": "Float",
    "Int": "Int",
    "Playlist": "Playlist",
    "Playlists": "Playlists",
    "Program": "Program",
    "ProgramEnrichment": "ProgramEnrichment",
    "ProgramSponsor": "ProgramSponsor",
    "Query": "Query",
    "RecordingMetadata": "RecordingMetadata",
    "String": "String",
    "__Directive": "__Directive",
    "__DirectiveLocation": "__DirectiveLocation",
    "__EnumValue": "__EnumValue",
    "__Field": "__Field",
    "__InputValue": "__InputValue",
    "__Schema": "__Schema",
    "__Type": "__Type",
    "__TypeKind": "__TypeKind",
  },
  "astNode": Object {
    "description": undefined,
    "directives": Array [],
    "kind": "SchemaDefinition",
    "loc": Object {
      "end": 23,
      "start": 0,
    },
    "operationTypes": Array [
      Object {
        "kind": "OperationTypeDefinition",
        "loc": Object {
          "end": 21,
          "start": 9,
        },
        "operation": "query",
        "type": Object {
          "kind": "NamedType",
          "loc": Object {
            "end": 21,
            "start": 16,
          },
          "name": Object {
            "kind": "Name",
            "loc": Object {
              "end": 21,
              "start": 16,
            },
            "value": "Query",
          },
        },
      },
    ],
  },
  "description": undefined,
  "extensionASTNodes": Array [],
  "extensions": undefined,
}
`;
